Add global components doc (#781)
PrestaShop/docs: src/content/1.7/development/components/global-components.md

---

> title: Global components

Reviewer: ```suggestion
title: Global JavaScript components
```
Remember we have PHP + JS in this project 😄 let's be accurate

weight: 70
---

# Global components ?

> It's a [new system](https://github.com/PrestaShop/PrestaShop/blob/develop/admin-dev/themes/new-theme/js/app/utils/init-components.js) mainly created for module developers, which allow you to use components we already use in the core without importing it.

Reviewer: Maybe we can mention somewhere the ADR 😊 ?

You are now able to use components we expose to the `window.prestashop.component` object without importing anything.

This object contains two attributes :

- `component` which contain some components you can instanciate
- `instance` which contains every instances of initiated components

It allow module developers to avoid importing path such as `../../../../admin-dev/themes/new-module/js/components/translatable-field` because it was making hard to add a CI in order to build the module.

## How to use ?

There is an event sent after the execution of `initPrestashopComponents` which you can stick on : `PSComponentsInitiated`.

If you have the `events` package, you'll be able to catch the event like this :

```js
EventEmitter.on('PSComponentsInitiated', () => {
  [...]
})
```

> You are able to custom a component like this :

Reviewer 2: did you mean "to use a custom component" ?

---
title: Global JavaScript components
weight: 70
---

> # Global JavaScript components ?

Reviewer 3: ```suggestion
# Global JavaScript components?
```

It's a [new system](https://github.com/PrestaShop/PrestaShop/blob/develop/admin-dev/themes/new-theme/js/app/utils/init-components.js) mainly created for module developers, which allow you to use components we already use in the core without importing it. [Here is the ADR](https://github.com/PrestaShop/ADR/blob/master/0009-expose-js-components-using-window-variable.md) where we decided what we should do to avoid this problem.

You are now able to use components we expose to the `window.prestashop.component` object without importing anything.

> This object contains two attributes :

Reviewer 3: ```suggestion
This object contains two attributes:
```

- `component` which contain some components you can instantiate;
- `instance` which contains every instances of initiated components.

> It allow module developers to avoid importing path such as `../../../../admin-dev/themes/new-module/js/components/translatable-field` because it was making hard to add a CI in order to build the module.

Reviewer 3: ```suggestion
It allows module developers to avoid importing path such as `../../../../admin-dev/themes/new-module/js/components/translatable-field` because it was making hard to add a CI in order to build the module.
```

allow's' right?

> ## How to use ?

Reviewer 3: ```suggestion
## How to use?
```

> There is an event sent after the execution of `initPrestashopComponents` which you can stick on : `PSComponentsInitiated`.

Reviewer 3: ```suggestion
There is an event sent after the execution of `initPrestashopComponents` which you can stick on: `PSComponentsInitiated`.
```

> If you have the `events` package, you'll be able to catch the event like this:

Reviewer 3: @NeOMakinG just for confirmation, `EventEmitter` is available globally? we can use `EventEmitter.on` or do we need to import `events`?

Author: Depends on which version you are, it's available globally only after 1.7.8

Reviewer 4: @NeOMakinG Could you add anot about that. And it's ok for me.

```js
EventEmitter.on('PSComponentsInitiated', () => {
  [...]
})
```

You are able to use a custom component [following this example](https://github.com/PrestaShop/example-modules/blob/master/demosymfonyform/views/js/form.js):

```js
$(document).ready(function() {
  window.prestashop.component.initComponents([
    'TranslatableField',
    'TinyMCEEditor',
    'TranslatableInput',
  ]);
});
```

You are also able to initiate it manually:

```js
EventEmitter.on('PSComponentsInitiated', () => {
  window.prestashop.component.TranslatableField = MyCustomTranslatableField;

  if (window.prestashop.component.instance.translatableField) {
    window.prestashop.component.instance.translatableField = new window.prestashop.component.TranslatableField();
  }
});
```

> or even import your own components that you want to make usable globally :

Reviewer 3: ```suggestion
or even import your own components that you want to make usable globally:
```

---
title: Global JavaScript components
weight: 70
---

# Global JavaScript components ?

> It's a [new system](https://github.com/PrestaShop/PrestaShop/blob/develop/admin-dev/themes/new-theme/js/app/utils/init-components.js) mainly created for module developers, which allow you to use components we already use in the core without importing it. [Here is the ADR](https://github.com/PrestaShop/ADR/blob/master/0009-expose-js-components-using-window-variable.md) we used to decide what we should do to avoid this problem.

Reviewer 4: ```suggestion
It's a [new system](https://github.com/PrestaShop/PrestaShop/blob/develop/admin-dev/themes/new-theme/js/app/utils/init-components.js) mainly created for module developers, which allow you to use components we already use in the core without importing it. [Here is the ADR](https://github.com/PrestaShop/ADR/blob/master/0009-expose-js-components-using-window-variable.md) where we decided what we should do to avoid this problem.
```

You are now able to use components we expose to the `window.prestashop.component` object without importing anything.

This object contains two attributes :

> - `component` which contain some components you can instanciate

Reviewer 4: ```suggestion
- `component` which contain some components you can instanciate;
```

> - `instance` which contains every instances of initiated components

Reviewer 4: ```suggestion
- `instance` which contains every instances of initiated components.
```

If you have the `events` package, you'll be able to catch the event like this :

```js
EventEmitter.on('PSComponentsInitiated', () => {
  [...]
})
```

You are able to use a custom component [following this example](https://github.com/PrestaShop/example-modules/blob/master/demosymfonyform/views/js/form.js):

```js
$(document).ready(function() {
  window.prestashop.component.initComponents([
    'TranslatableField',
    'TinyMCEEditor',
    'TranslatableInput',
  ]);
});
```

You are also able to initiate it manually:

```js
EventEmitter.on('PSComponentsInitiated', () => {
  window.prestashop.component.TranslatableField = MyCustomTranslatableField;

  if (window.prestashop.component.instance.translatableField) {
    window.prestashop.component.instance.translatableField = new window.prestashop.component.TranslatableField();
  }
});
```

or even import your own components that you want to make usable globally :

```js
EventEmitter.on('PSComponentsInitiated', () => {
  window.prestashop.component.MyCustomComponent = MyCustomComponent;
  window.prestashop.component.instance.myCustomComponent = new window.prestashop.component.MyCustomComponent();
});
```

> ## What if components are not initiated ?

Reviewer 4: Question not answered
